master/buildbot/util/kubeclientservice.py

```diff
@@ -0,0 +1,249 @@
+import abc
+import os
+import time
+
+from twisted.internet import defer
+from twisted.internet import reactor
+from twisted.internet.error import ProcessExitedAlready
+from twisted.internet.protocol import ProcessProtocol
+from twisted.python.failure import Failure
+
+from buildbot import config
+from buildbot.compat import TimeoutError
+from buildbot.util import asyncSleep
+from buildbot.util.httpclientservice import HTTPClientService
+from buildbot.util.logger import Logger
+from buildbot.util.service import BuildbotService
+
+log = Logger()
+
+
+# this is a BuildbotService, so that it can be started and destroyed.
+# this is needed to implement kubectl proxy lifecycle
+class KubeConfigLoaderBase(BuildbotService):
+    name = "KubeConfig"
+
+    @abc.abstractmethod
+    def getConfig(self):
+        """
+        @return dictionary with optional params
+        {
+            'master_url': 'https://kube_master.url',
+            'namespace': 'default_namespace',
+            'headers' {
+                'Authentication': XXX
+            }
+            # todo (quite hard to implement with treq):
+            'cert': 'optional client certificate used to connect to ssl'
+            'verify': 'kube master certificate authority to use to connect'
+        }
+        """
+
+    def __str__(self):
+        """return unique str for SharedService"""
+        # hash is implemented from ComparableMixin
+        return "{}({})".format(self.__class__.__name__, hash(self))
+
+
+class KubeHardcodedConfig(KubeConfigLoaderBase):
+    def reconfigService(self,
+                        master_url=None,
+                        headers=None,
+                        cert=None,
+                        verify=None,
+                        namespace="default"):
+        self.config = {'master_url': master_url, 'namespace': namespace}
+        if headers is not None:
+            self.config['headers'] = headers
+        if cert is not None:
+            self.config['cert'] = cert
+        if verify is not None:
+            self.config['verify'] = verify
+
+    checkConfig = reconfigService
+
+    def getConfig(self):
+        return self.config
+
+
+class KubeCtlProxyConfigLoader(KubeConfigLoaderBase):
+    """ We use kubectl proxy to connect to kube master.
+    Parsing the config and setting up SSL is complex.
+    So for now, we use kubectl proxy to load the config and connect to master.
+    This will run the kube proxy as a subprocess, and return configuration with
+    http://localhost:PORT
+    """
+    kube_ctl_proxy_cmd = ['kubectl', 'proxy']  # for tests override
+
+    class LocalPP(ProcessProtocol):
+        def __init__(self, config_loader):
+            self.config_loader = config_loader
+            self.got_output_deferred = defer.Deferred()
+            self.terminated_deferred = defer.Deferred()
+
+        def outReceived(self, data):
+            if not self.got_output_deferred.called:
+                self.got_output_deferred.callback(data)
```

Comment:
Do we have any guarantees of how chunked input to `outReceived` is? We could get a single character worst case, couldn't we? I think ﻿`utils.getProcessOutputAndValue` would be better here.

Reply:
Maybe we should wait for the first line break and call `got_output_deferred` only then?